packages/backend/src/open_payments/payment/outgoing/service.ts

```diff
@@ -159,6 +185,326 @@ async function createOutgoingPayment(
+  access,
+  payment,
+  identifier
+}: {
+  access: GrantAccess
+  payment: OutgoingPayment
+  identifier: string
+}): boolean {
+  if (
+    (!access.identifier || access.identifier === identifier) &&
+    (!access.interval || !!getInterval(access.interval, payment.createdAt)) &&
+    (!access.limits || validateAccessLimits(payment, access.limits))
+  ) {
+    return true
+  }
+  return false
+}
+
+function validatePaymentAccess({
+  access,
+  payment,
+  identifier
+}: {
+  access: PaymentAccess
+  payment: OutgoingPayment
+  identifier: string
+}): boolean {
+  if (
+    (!access.identifier || access.identifier === identifier) &&
+    (!access.paymentInterval ||
+      (access.paymentInterval.start.toMillis() <= payment.createdAt.getTime() &&
+        payment.createdAt.getTime() < access.paymentInterval.end.toMillis())) &&
+    (!access.limits || validateAccessLimits(payment, access.limits))
+  ) {
+    return true
+  }
+  return false
+}
+
+function validateAccessLimits(
+  payment: OutgoingPayment,
+  limits: AccessLimits
+): boolean {
+  if (
+    (!limits.receiver || payment.receiver === limits?.receiver) &&
+    (!limits.description || payment.description === limits?.description) &&
+    (!limits.externalRef || payment.externalRef === limits?.externalRef) &&
+    validateAmountAssets(payment, limits)
+  ) {
+    return true
+  }
+  return false
+}
+
+function validateAmountAssets(
+  payment: OutgoingPayment,
+  limits: AccessLimits
+): boolean {
+  if (
+    limits.sendAmount &&
+    (limits.sendAmount.assetCode !== payment.asset.code ||
+      limits.sendAmount.assetScale !== payment.asset.scale)
+  ) {
+    return false
+  }
+  return (
+    !limits.receiveAmount ||
+    (limits.receiveAmount.assetCode === payment.receiveAmount.assetCode &&
+      limits.receiveAmount.assetScale === payment.receiveAmount.assetScale)
+  )
+}
+
+function estimateTotalAvailableAmounts<T extends GrantAccess>(
+  accessArray: T[],
+  payment: OutgoingPayment,
+  slippage: number
+) {
+  const availableSendAmount = accessArray.reduce(
+    (prev: bigint, access: T) =>
+      prev + (access.limits?.sendAmount?.value ?? BigInt(0)),
+    BigInt(0)
+  )
+
+  const availableReceiveAmount = accessArray.reduce(
+    (prev: bigint, access: T) =>
+      prev + (access.limits?.receiveAmount?.value ?? BigInt(0)),
+    BigInt(0)
+  )
+
+  const estTotalAvailableSendAmount =
+    availableSendAmount +
+    BigInt(
+      Math.floor(
+        (Number(payment.quote.lowEstimatedExchangeRate.b) /
+          Number(payment.quote.lowEstimatedExchangeRate.a)) *
+          (1 + slippage) *
+          Number(availableReceiveAmount)
+      )
+    )
+  const estTotalAvailableReceiveAmount =
+    availableReceiveAmount +
+    BigInt(
+      Math.floor(
+        (Number(payment.quote.lowEstimatedExchangeRate.a) /
+          Number(payment.quote.lowEstimatedExchangeRate.b)) *
+          (1 + slippage) *
+          Number(availableSendAmount)
+      )
+    )
```

Comment:
It's not clear to me why we wouldn't just use `availableSendAmount` and `availableReceiveAmount`

Reply:
I probably don't need to calculate both. I think we should work through this part together next week.